Mathlib/Algebra/Hom/Group.lean

```diff
@@ -0,0 +1,1694 @@
+  ↑(ofMapDiv f hf) = f := rfl
+#align monoid_hom.coe_of_map_div MonoidHom.coe_of_map_div
+
+/-- If `f` is a monoid homomorphism to a commutative group, then `f⁻¹` is the homomorphism sending
+`x` to `(f x)⁻¹`. -/
+@[to_additive]
+instance {M G} [MulOneClass M] [CommGroup G] : Inv (M →* G) :=
+  ⟨fun f => (mk' fun g => (f g)⁻¹) fun a b => by rw [← mul_inv, f.map_mul]⟩
+
+/-- If `f` is an additive monoid homomorphism to an additive commutative group, then `-f` is the
+homomorphism sending `x` to `-(f x)`. -/
+add_decl_doc AddMonoidHom.instNegAddMonoidHomToAddZeroClassToAddMonoidToSubNegAddMonoidToAddGroup
+
+@[simp, to_additive]
+theorem inv_apply {M G} {_ : MulOneClass M} {_ : CommGroup G} (f : M →* G) (x : M) :
+  f⁻¹ x = (f x)⁻¹ := rfl
+#align monoid_hom.inv_apply MonoidHom.inv_apply
+
+@[simp, to_additive]
+theorem inv_comp {M N A} {_ : MulOneClass M} {_ : MulOneClass N} {_ : CommGroup A}
+  (φ : N →* A) (ψ : M →* N) : φ⁻¹.comp ψ = (φ.comp ψ)⁻¹ := by
+  ext
+  simp only [Function.comp_apply, inv_apply, coe_comp]
+#align monoid_hom.inv_comp MonoidHom.inv_comp
+
+@[simp, to_additive]
+theorem comp_inv {M A B} {_ : MulOneClass M} {_ : CommGroup A} {_ : CommGroup B}
+  (φ : A →* B) (ψ : M →* A) : φ.comp ψ⁻¹ = (φ.comp ψ)⁻¹ := by
+  ext
+  simp only [Function.comp_apply, inv_apply, map_inv, coe_comp]
+#align monoid_hom.comp_inv MonoidHom.comp_inv
+
+/-- If `f` and `g` are monoid homomorphisms to a commutative group, then `f / g` is the homomorphism
+sending `x` to `(f x) / (g x)`. -/
+@[to_additive]
+instance {M G} [MulOneClass M] [CommGroup G] : Div (M →* G) :=
+  ⟨fun f g => (mk' fun x => f x / g x) fun a b => by
+    simp [div_eq_mul_inv, mul_assoc, mul_left_comm, mul_comm]⟩
+
+/-- If `f` and `g` are monoid homomorphisms to an additive commutative group, then `f - g`
+is the homomorphism sending `x` to `(f x) - (g x)`. -/
+add_decl_doc AddMonoidHom.instSubAddMonoidHomToAddZeroClassToAddMonoidToSubNegAddMonoidToAddGroup
+
+@[simp, to_additive]
+theorem div_apply {M G} {_ : MulOneClass M} {_ : CommGroup G} (f g : M →* G) (x : M) :
+  (f / g) x = f x / g x := rfl
+#align monoid_hom.div_apply MonoidHom.div_apply
+
+end MonoidHom
+
+-- Porting note: delete once `to_additive` in `MonoidHom.map_zpow` is fixed
+/-- Additive group homomorphisms preserve integer scaling. -/
+protected theorem AddMonoidHom.map_zsmul [AddGroup α] [SubtractionMonoid β]
+  (f : α →+ β) (g : α) (n : ℤ) :
+  f (n • g) = n • f g := map_zsmul f n g
+#align add_monoid_hom.map_zsmul AddMonoidHom.map_zsmul
+
+/-- Given two monoid with zero morphisms `f`, `g` to a commutative monoid, `f * g` is the monoid
+with zero morphism sending `x` to `f x * g x`. -/
+instance {M N} {_ : MulZeroOneClass M} [CommMonoidWithZero N] : Mul (M →*₀ N) :=
+  ⟨fun f g => { (f * g : M →* N) with
+    toFun := fun a => f a * g a,
+    map_zero' := by dsimp only []; rw [map_zero, zero_mul] }⟩
+    -- Porting note: why do we need `dsimp` here?
+
+section coe
+
+/-! ### Coercions as bundled morphisms
```

Comment:
These classes and theorems may be be wrong approach in Lean4 due to eager elaboration of coercions. I'm not confident enough to refactor / remove this section.
https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Coe.2C.20CoeHead.2C.20CoeTail/near/312411053

Reply:
We're attempting the revert now. I guess this PR will have to wait for that.

Reply:
The revert is now in. Let's revise this PR to match mathlib3 again.

Reply:
@winstonyin, this can be resolved now, right?